src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver.cc

```diff
@@ -185,6 +194,14 @@ static fd_node* pop_fd_node_locked(fd_node** head, ares_socket_t as) {
   return nullptr;
 }
 
+static void on_timeout_locked(void* arg, grpc_error* error) {
+  grpc_ares_ev_driver* driver = static_cast<grpc_ares_ev_driver*>(arg);
+  if (error == GRPC_ERROR_NONE) {
```

Comment:
What happens if this closure is already scheduled with `GRPC_ERROR_NONE`, but before it actually runs, the driver completes and is shut down?  In that case, we need to avoid calling `grpc_ares_ev_driver_shutdown_locked()` again, don't we?

Reply:
this was originally relying on the idempotency `driver_shutdown_locked`, in that calling it a second time is effectively a no-op. But that said I do think it's cleaner to be more explicit here, changed this to avoid do anything in the timer if we've already reached `driver->shutting_down = true` state.